src/events/processSignatureRequests.js

```diff
@@ -15,6 +15,8 @@ function processSignatureRequestsBuilder(config) {
   return async function processSignatureRequests(signatureRequests) {
     const txToSend = []
 
+    const expectedMessageLength = await homeBridge.methods.requiredMessageLength().call()
```

Comment:
I agree that this is not ideal, but notice that one such request will be done for each _group_ of signature requests.

The problem is that I don't see a proper place to do this that won't require a big change or reduce the quality of the code:

1. Doing it as part of `processSignatureRequestBuilder` would make that function asynchronous and we would need to change how it's used (and the other builders too, if we want to be consistent).
2. Same issue if we make the requiredMessageLength a part of the received config: we get the configuration synchronously now, so we'd need some non-trivial changes.
3. It could be part of the watcher's `initialize` function, but this is only required for one watcher, so we'd need to pass it to all functions unnecessarily or have a conditional. Besides, we don't have access to the contract there (and it wouldn't make sense to have the home bridge contract in the watcher for the foreign bridge contract).

I'm not saying that this can't be done, just that we have a trade-off. In my opinion, the best options are leaving it as it is (I don't think it will be a huge performance penalty), or making it part of the config (option 2) and making all configurations asynchronous (which maybe we'll need to do later anyway).

Reply:
Could you measure handling of 300 txs - average time between the moment when the event was observed and the moment when `submitSignature()` is called. One test with new changes and another test without changes. We could see difference.

Reply:
We can't do that test using what we have now, because the method call is done before the transaction log. What I did was to move the call _inside_ the loop and test that (much worse) scenario.

Mean with the call: **4658 ms**
Mean without the call: **4159 ms**

This was measured with a block of 200 txs. 501 ms are added. Divided by the number of transactions (200), we have 2.5 extra ms per transaction. But, since the call will actually be outside the loop, that's just 2.5 ms per block.

Take into account that I used a local chain. If you are using a remote RPC, this overhead will be bigger.

Reply:
I think @akolotov is right, we shouldn't do it every single time. Maybe do it like every 1000 tx?

Reply:
thanks @rstormsf. It is a good idea! I think better idea could be to use the singleton template to read this value once the sender is started?

Reply:
Could you clarify why you said that it is used in all watchers? My understanding is that  `processSignatureRequests` is already used only in one specific watcher:
https://github.com/poanetwork/bridge-nodejs/blob/c24917a2508348c371c00bf73a4fa19d0f6a1f4d/src/watcher.js#L77-L78

Reply:
Yes, I was referring to the `watcher` code, that is shared by all watchers. What did you mean with "when is started"?

Reply:
Sorry, for confusion. I should say 'watcher' that time. 

I meant the code which could look like this:

```
  return async function processSignatureRequests(signatureRequests) {
    const txToSend = []

    if ( expectedMessageLength === undefined ) {
      const expectedMessageLength = await homeBridge.methods.requiredMessageLength().call()
    }

    const callbacks = signatureRequests.map(async signatureRequest => {
      const { recipient, value } = signatureRequest.returnValues
```

Reply:
Oh, I see! That could work, sure. The only caveat is that if the message length changes, it won't be updated. But that can be fixed by restarting the bridge.

Reply:
Cool! It doesn't mean to be updated often as I said. So, restart of the bridge is suitable.

Reply:
wait...
`expectedMessageLength` will always be undefined, so we will always query RPC every single time

Reply:
@fvictorio tonught I have been thinking about another possible solution...
What if try to re-read the message length every time when this assert raises an exception:
https://github.com/poanetwork/bridge-nodejs/blob/3a127ff8fbe742e920ad7608460453e7f43ab0b9/src/utils/message.js#L32

So, the logic could be:
1. If `expectedMessageLength` is not defined, request the contract for the value.
2. if `createMessage` fails due to incorrect `expectedMessageLength` re-read a new value from contract and invoke `createMessage` once again. If it still fails raise an exception.